docs/user-guide/migrating-to-4.0.0.md

```diff
@@ -167,7 +167,15 @@ Prior to 4.0, ESLint required parsers to implement comment attachment, a process
 
 In 4.0, comment attachment logic has been moved into ESLint itself. This should make it easier to develop custom parsers, but it also means that AST nodes will no longer have `leadingComments` and `trailingComments` properties. Additionally, `LineComment` and `BlockComment` events will no longer be emitted during AST traversal.
 
-**To address:** If you have a custom rule that depends on the `leadingComments` or `trailingComments` properties of an AST node, you can switch to `sourceCode.getComments(node).leading` or `sourceCode.getComments(node).trailing` instead. You should also consider using `sourceCode.getAllComments()` or `sourceCode.getTokenBefore(node, { includeComments: true })`.
+**To address:** If you have a custom rule that depends on the `leadingComments` or `trailingComments` properties of an AST node, you can now use `sourceCode.getCommentsBefore()` and `sourceCode.getCommentsAfter()` instead, respectively.
```

Comment:
Just realizing there doesn't seem to be a clear answer for when `LineComment` and `BlockComment` events are no longer emitted. Not directly related to your PR, so maybe one of us should create a separate issue.